File: src/main/resources/META-INF/modules/exanpe/gMap.js
/*
 * Copyright 2013 EXANPE <exanpe@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

define(["exanpe/common",
		"yui/2.9.0/yahoo-dom-event", 
		"yui/2.9.0/json-min"], function(gMap) {
	
	/** 
	 * Constructor
	 * @class Represents a GoogleMap component. Can be accessed through JavaScript by its id.
	 * @param {String} id the id of the component
	 * @param {String} latitude the latitude position used to center the map
	 * @param {String} longitude the longitude position used to center the map 
	 * @param {String} mapType the type of map to display
	 * @param {int} zoom the initial Map zoom level
	 * @param {boolean} draggable indicating if the map is draggable or not
	 * @param {Object[]} markers the markers to display into the map
	 * @param {Object[]} polyPoints the points used to display polyline into the map
	 * @param {String} polyStrokeColor the color of Polyline/Polygon lines
	 * @param {String} polyStrokeOpacity the opacity Polyline/Polygon line's color
	 * @param {String} polyStrokeWeight the weight of Polyline/Polygo line's in pixels.
	 * @param {boolean} polygon draw a Polygon instead of Polyline (default) if set to true
	 */
	Exanpe.GMap = function(id, latitude, longitude, mapType, zoom, draggable, markers, polyPoints, polyStrokeColor, polyStrokeOpacity, polyStrokeWeight, polygon){
		
		/**
		 * Id of the instance
		 */
		this.id = id;
		
		/**
		 * Initial Latitude used to center the map
		 */
		this.latitude = latitude;
		
		/**
		 * Initial Longitude used to center the map
		 */
		this.longitude = longitude;
		
		/**
		 * The type of map to display (ROADMAP, HYBRID, ...)
		 */
		this.mapType = mapType;
		
		/**
		 * Initial Map zoom level
		 */
		this.zoom = zoom;
		
		/**
		 * If the map is draggable or not
		 */
		this.draggable = draggable;
		
		/**
		 * Google Map options
		 */
		this.options = null;
		
		/**
		 * Google Map info window object
		 */
		this.infoWindow = null;
		
		/**
		 * Google Map markers to display
		 */
		this.markers = markers;
		
		/**
		 * Google Map points used by Polyline
		 */
		this.polyPoints = polyPoints;
		
		/**
		 * Color of Polyline/Polygon lines
		 */
		this.polyStrokeColor = polyStrokeColor;
		
		/**
		 * Opacity of Polyline/Polygon line's color
		 */
		this.polyStrokeOpacity = polyStrokeOpacity;
		
		/**
		 * Weight of Polyline/Polygo line's in pixels.
		 */
		this.polyStrokeWeight = polyStrokeWeight;
		
		/**
		 * Draw a Polygon instead of Polyline (default) if set to true
		 */
		this.polygon = polygon;
		
		/**
		 * Google Map object
		 */
		this.map = null;
		
		/**
		 * Google Map Poly object
		 */
		this.poly = null;
	};

	/**
	 * GoogleMap prefix
	 * @constant
	 * @static
	 * @private
	 */
	Exanpe.GMap.GOOGLE_MAP_PREFIX = "exanpe-gmap-";

	/**
	 * Dom method utility
	 * @private
	 */
	Exanpe.GMap.prototype.getMapContainerEl = function() {
		return YAHOO.util.Dom.get(Exanpe.GMap.GOOGLE_MAP_PREFIX + this.id);
	};

	/** 
	 * Initializes the Map object
	 * @private
	 */
	Exanpe.GMap.prototype._initMap = function(){
		// Init windows
		this.infoWindow = new google.maps.InfoWindow();
		
		// Options
		var position = new google.maps.LatLng(this.latitude, this.longitude);
		this.options = {
				center: position,
				zoom: this.zoom,					
				draggable: this.draggable,
				mapTypeId: eval("google.maps.MapTypeId." + this.mapType),
				mapTypeControl: true,
				mapTypeControlOptions: {
				  style: google.maps.MapTypeControlStyle.DROPDOWN_MENU
				},					
				navigationControl: true,
				navigationControlOptions: {
				  style: google.maps.NavigationControlStyle.SMALL
				}
			};		
		
		// Init map
		this.map = new google.maps.Map(this.getMapContainerEl(), this.options);
	}

	/** 
	 * Initializes the Poly object
	 * @private
	 */
	Exanpe.GMap.prototype._initPoly = function(){
		  var polyOptions = {
				    strokeColor: this.polyStrokeColor,
				    strokeOpacity: this.polyStrokeOpacity,
				    strokeWeight: this.polyStrokeWeight
		  }
		  
		  if (this.polygon) {
			  this.poly = new google.maps.Polygon(polyOptions);
		  }
		  else {
			  this.poly = new google.maps.Polyline(polyOptions);  
		  }
		  
		  this.poly.setMap(this.map);
	};

	/** 
	 * Initializes the GMap Markers
	 * @private
	 */
	Exanpe.GMap.prototype._initMarker = function(marker) {
		// Add Marker
		var markerPosition = new google.maps.LatLng(marker.latitude, marker.longitude);
		var gmarker = new google.maps.Marker({
				position: markerPosition,
				map: this.map,
				icon : marker.icon
			}
		);	
		
		// Add Marker event
		var info = null
		var map = this.map;
		var iw = this.infoWindow;
		if (marker.info) {
			info = marker.info;
		}
		
		// Marker listener
		google.maps.event.addListener(gmarker, 'click', function() {
				if (info) {
					iw.setContent(info);
				}
				else {
					iw.close();
					iw.setContent("");
				}
				iw.open(map, this);
				map.panTo(markerPosition);
			}
		);
		
		// Marker link event
		var mapItem = YAHOO.util.Dom.get(marker.id);
		YAHOO.util.Event.addListener(mapItem, "click", function() {
			if (info) {
				iw.setContent(info);
			}
			else {
				iw.close();
				iw.setContent("");
			}
			iw.open(map, gmarker);
			map.panTo(markerPosition);
		});
		
	};

	/** 
	 * Initializes the Poly object
	 * @private
	 */
	Exanpe.GMap.prototype._initPolyPoint = function(point) {
		// Add new Point to the poly path
		var pointPosition = new google.maps.LatLng(point.latitude, point.longitude);
		var path = this.poly.getPath();
		path.push(pointPosition);
	};

	/** 
	 * Prepare map and items to display
	 * @private
	 */
	Exanpe.GMap.prototype._init = function(){
		// Init map
		this._initMap();

		// Init markers
		for (var i = 0; i < this.markers.length; i++) {
			var marker = this.markers[i];
			this._initMarker(marker);
		}
		
		// Init Polyline / Polygon
		if (this.polyPoints.length > 0) {
			this._initPoly();
		}
		for (var i = 0; i < this.polyPoints.length; i++) {
			var point = this.polyPoints[i];
			this._initPolyPoint(point);
		}
	};

	/**
	 * Initializes the GoogleMap component on DOM load
	 * @param {Object} data the json data coming from Java class initialization
	 * @private
	 * @static
	 */
	function gMapBuilder(data) {
		var gMap = new Exanpe.GMap(data.id, data.latitude, data.longitude, data.mapType, data.zoom, data.draggable===true, YAHOO.lang.JSON.parse(data.markers), YAHOO.lang.JSON.parse(data.polyPoints), data.polyStrokeColor, data.polyStrokeOpacity, data.polyStrokeWeight, data.polygon===true);
		gMap._init();
		window[data.id] = gMap;
	};
	
	return {
		init: gMapBuilder
	}
	
});

File: src/main/java/fr/exanpe/t5/lib/components/GMap.java
//
// Copyright 2011 EXANPE <exanpe@gmail.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package fr.exanpe.t5.lib.components;

import org.apache.commons.lang.StringUtils;
import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.ComponentResources;
import org.apache.tapestry5.MarkupWriter;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.BeginRender;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.annotations.SetupRender;
import org.apache.tapestry5.annotations.SupportsInformalParameters;
import org.apache.tapestry5.dom.Element;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.ioc.annotations.Symbol;
import org.apache.tapestry5.json.JSONArray;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.Environment;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;
import org.slf4j.Logger;

import fr.exanpe.t5.lib.constants.ExanpeSymbols;
import fr.exanpe.t5.lib.constants.GMapTypeEnum;
import fr.exanpe.t5.lib.model.GMapInternalModel;
import fr.exanpe.t5.lib.model.gmap.GMapMarkerModel;
import fr.exanpe.t5.lib.model.gmap.GMapPolyPointModel;
import fr.exanpe.t5.lib.services.ExanpeComponentService;

/**
 * GoogleMap component.<br />
 * Displays a Google Map centered with latitude and longitude positions provided by user.<br />
 * {@link GMapMarker} and/or {@link GMapPolyPoint} components can be nested in its body in order to
 * display custom markers or draw polylines or polygons.<br />
 * <p>
 * JavaScript : This component is bound to a class Exanpe.GMap.<br/>
 * CSS : This component is bound to a class exanpe-gmap.<br/>
 * </p>
 * 
 * @since 1.2
 * @see GMapMarker
 * @see GMapPolyPoint
 * @author lguerin
 */
@Import(stylesheet =
{ "css/exanpe-t5-lib-core.css", "css/exanpe-t5-lib-skin.css" })
@SupportsInformalParameters
public class GMap implements ClientElement
{
    /**
     * Initial latitude used to center the map.
     */
    @Property
    @Parameter(required = true, allowNull = false, defaultPrefix = BindingConstants.LITERAL)
    private String latitude;

    /**
     * Initial longitude used to center the map.
     */
    @Property
    @Parameter(required = true, allowNull = false, defaultPrefix = BindingConstants.LITERAL)
    private String longitude;

    /**
     * Specify the type of map you want to display. Default is ROADMAP.
     * 
     * @see GMapTypeEnum
     */
    @Parameter(value = "ROADMAP", allowNull = false, defaultPrefix = BindingConstants.LITERAL)
    private GMapTypeEnum mapType;

    /**
     * Initial Map zoom level
     */
    @Parameter(value = "15", allowNull = false, defaultPrefix = BindingConstants.LITERAL)
    private Integer zoom;

    /**
     * If false, prevents the map from being dragged. Default value is true.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "true")
    private Boolean draggable;

    /**
     * If true, displays a side bar with marker links. Default is false.
     */
    @Property
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "false")
    private Boolean sidebar;

    /**
     * Used with {@link GMapPolyPoint} component in order to specify the hexadecimal HTML color of
     * a line.
     * Default is black (#000);
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "#000")
    private String polyStrokeColor;

    /**
     * Used with {@link GMapPolyPoint} component in order to specify numerical fractional value
     * between 0.0 and 1.0 of the opacity of the line's color.
     * Default is 1.0
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "1.0")
    private float polyStrokeOpacity;

    /**
     * Used with {@link GMapPolyPoint} component in order to specify the weight of the line's
     * stroke in pixels.
     * Default is 3.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "3")
    private Integer polyStrokeWeight;

    /**
     * Used with {@link GMapPolyPoint} component in order to draw a Javascript GMap Polygon instead
     * of a Polyline.
     * Default is false.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "false")
    private Boolean polygon;

    /**
     * Loading Google Maps API over HTTPS or not.
     * Default is false.
     */
    @Parameter(defaultPrefix = BindingConstants.LITERAL, value = "false")
    private Boolean secure;

    @Property
    private GMapInternalModel model;

    /**
     * Current marker looped by model
     */
    @Property
    private GMapMarkerModel current;

    @Property
    private String uniqueId;

    /**
     * Root CSS class
     */
    private static final String ROOT_CSS_CLASS = "exanpe-gmap";

    /**
     * Google Maps API
     */
    private static final String GOOGLE_MAP_API_BASE_URL = "http://maps.googleapis.com/maps/api/js?";

    @Inject
    @Symbol(ExanpeSymbols.GMAP_V3_BUSINESS_CLIENT_ID)
    private String gmapClientId;

    @Inject
    @Symbol(ExanpeSymbols.GMAP_V3_VERSION)
    private String gmapApiVersion;

    @Inject
    private ComponentResources resources;

    @Inject
    private Environment environment;

    @Inject
    private JavaScriptSupport javascriptSupport;

    @Inject
    private ExanpeComponentService ecservice;

    @Inject
    private Logger log;

    @SetupRender
    void init()
    {
        uniqueId = javascriptSupport.allocateClientId(resources);
        model = new GMapInternalModel();
        model.setMapId(uniqueId);

        if (environment.push(GMapInternalModel.class, model) != null) { throw new IllegalStateException("Nested GMap components are not supported"); }
    }

    @BeginRender
    void begin(MarkupWriter writer)
    {
        Element e = writer.element("div");
        e.attribute("id", getClientId());
        resources.renderInformalParameters(writer);
        ecservice.reorderCSSClassDeclaration(e, ROOT_CSS_CLASS);
    }

    @AfterRender
    void end(MarkupWriter writer)
    {
        writer.end();
        environment.pop(GMapInternalModel.class);
        JSONObject data = buildJSONData(model);
        javascriptSupport.importJavaScriptLibrary(this.buildGMapApiUrl());
        javascriptSupport.require("exanpe/gMap").invoke("init").with(data);
    }

    private JSONObject buildJSONData(GMapInternalModel model)
    {
        // Markers
        JSONArray gMapMarkers = buildJSONMarkersArray();

        // Polyline points
        JSONArray gMapPolyPoints = buildJSONPolyPointsArray();

        JSONObject data = new JSONObject();
        data.accumulate("id", getClientId());
        data.accumulate("latitude", latitude);
        data.accumulate("longitude", longitude);
        data.accumulate("mapType", mapType.toString());
        data.accumulate("zoom", zoom);
        data.accumulate("draggable", draggable);
        data.accumulate("markers", gMapMarkers.toString());
        data.accumulate("polyPoints", gMapPolyPoints.toString());
        data.accumulate("polyStrokeColor", polyStrokeColor);
        data.accumulate("polyStrokeOpacity", polyStrokeOpacity);
        data.accumulate("polyStrokeWeight", polyStrokeWeight);
        data.accumulate("polygon", polygon);
        return data;
    }

    private JSONArray buildJSONMarkersArray()
    {
        JSONArray gMapMarkers = new JSONArray();
        for (GMapMarkerModel m : model.getMarkers())
        {
            JSONObject marker = new JSONObject();
            marker.accumulate("id", m.getId());
            marker.accumulate("icon", m.getIcon() != null ? m.getIcon().toClientURL() : "");
            marker.accumulate("latitude", m.getLatitude());
            marker.accumulate("longitude", m.getLongitude());
            marker.accumulate("title", m.getTitle());
            marker.accumulate("info", m.getInfo());
            gMapMarkers.put(marker);
        }
        return gMapMarkers;
    }

    private JSONArray buildJSONPolyPointsArray()
    {
        JSONArray gMapPolyPoints = new JSONArray();
        for (GMapPolyPointModel m : model.getPolyPoints())
        {
            JSONObject point = new JSONObject();
            point.accumulate("id", m.getId());
            point.accumulate("latitude", m.getLatitude());
            point.accumulate("longitude", m.getLongitude());
            gMapPolyPoints.put(point);
        }
        return gMapPolyPoints;
    }

    private String buildGMapApiUrl()
    {
        String url = GOOGLE_MAP_API_BASE_URL;

        // GMap version
        if (!gmapApiVersion.matches("^3(\\.[0-9]{1})?$")) { throw new RuntimeException(
                String.format("The GMap API version: %s is not correct.", gmapApiVersion)); }
        url += "v=" + gmapApiVersion;

        // GMap business client ID
        if (!StringUtils.isEmpty(gmapClientId))
        {
            if (!StringUtils.startsWith(gmapClientId, "gme-")) { throw new RuntimeException(String.format(
                    "Your GMap clientID: %s is not correct, it must start with 'gme-'.",
                    gmapClientId)); }
            url += "&client=" + gmapClientId;
        }

        // Sensor parameter
        url += "&sensor=true";

        // Secure access
        if (secure)
        {
            url = url.replace("http", "https");
        }

        log.debug("Google Maps API URL: {}", url);
        return url;
    }

    public String getClientId()
    {
        return uniqueId;
    }
}
